File: examples/apps/riffer/go/clips_bridge.go
//go:build nxuskit

// Package riffer provides music sequence analysis and transformation.
//
// CLIPS Rule Engine Bridge for Riffer.
// Provides integration with the CLIPS expert system for scoring adjustments
// and context-aware suggestions.
package riffer

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"time"

	nxuskit "github.com/nxus-SYSTEMS/nxusKit/packages/nxuskit-go"
)

// ScoringAdjustment represents a scoring adjustment from CLIPS rules
type ScoringAdjustment struct {
	Dimension  string  `json:"dimension"`
	Adjustment float32 `json:"adjustment"`
	Reason     string  `json:"reason"`
	RuleName   string  `json:"rule_name"`
}

// ClipsSuggestion represents a suggestion from CLIPS rules
type ClipsSuggestion struct {
	Category    string `json:"category"`
	Severity    string `json:"severity"`
	Message     string `json:"message"`
	RuleName    string `json:"rule_name"`
	NoteIndices []int  `json:"note_indices,omitempty"`
}

// ClipsResult contains the results from CLIPS rule execution
type ClipsResult struct {
	Adjustments     []ScoringAdjustment `json:"adjustments"`
	Suggestions     []ClipsSuggestion   `json:"suggestions"`
	RulesFired      uint32              `json:"rules_fired"`
	ExecutionTimeMs uint64              `json:"execution_time_ms"`
}

// ClipsRuleEngine provides CLIPS integration for music analysis
type ClipsRuleEngine struct {
	rulesDir  string
	available bool
}

// NewClipsRuleEngine creates a new CLIPS rule engine
func NewClipsRuleEngine(rulesDir string) (*ClipsRuleEngine, error) {
	// Check if rules directory exists
	info, err := os.Stat(rulesDir)
	available := err == nil && info.IsDir()

	if !available {
		fmt.Fprintf(os.Stderr, "Warning: CLIPS rules directory not found at %s. Using deterministic scoring only.\n", rulesDir)
	}

	return &ClipsRuleEngine{
		rulesDir:  rulesDir,
		available: available,
	}, nil
}

// IsAvailable returns true if CLIPS is available
func (e *ClipsRuleEngine) IsAvailable() bool {
	return e.available
}

// SequenceToFacts converts a sequence to CLIPS facts JSON
func (e *ClipsRuleEngine) SequenceToFacts(sequence *Sequence) map[string]interface{} {
	facts := make([]map[string]interface{}, 0)

	// Add context fact
	contextFact := e.contextToFact(&sequence.Context)
	facts = append(facts, contextFact)

	// Add note facts
	for idx, note := range sequence.Notes {
		noteFact := e.noteToFact(&note, idx)
		facts = append(facts, noteFact)
	}

	// Add interval facts
	intervals := AnalyzeIntervals(sequence.Notes)
	for _, interval := range intervals {
		intervalFact := e.intervalToFact(&interval)
		facts = append(facts, intervalFact)
	}

	// Add scale membership facts
	if sequence.Context.KeySignature != nil {
		key := NewKeySignature(sequence.Context.KeySignature.Root, sequence.Context.KeySignature.Mode)
		for idx, note := range sequence.Notes {
			pc := PitchClassFromMidi(note.Pitch)
			inScale := IsInScale(pc, key)
			membershipFact := map[string]interface{}{
				"template": "scale-membership",
				"values": map[string]interface{}{
					"note-index":  idx,
					"pitch-class": pc.String(),
					"in-scale":    inScale,
				},
			}
			facts = append(facts, membershipFact)
		}
	}

	// Add dissonance facts for strong dissonances
	for _, interval := range intervals {
		if interval.Quality == StrongDissonance {
			dissonanceFact := map[string]interface{}{
				"template": "dissonance",
				"values": map[string]interface{}{
					"note-index":         interval.FromIndex,
					"next-index":         interval.ToIndex,
					"interval-semitones": interval.Semitones,
					"needs-resolution":   true,
				},
			}
			facts = append(facts, dissonanceFact)
		}
	}

	return map[string]interface{}{
		"facts": facts,
		"config": map[string]interface{}{
			"include_trace": true,
			"max_rules":     100,
		},
	}
}

// contextToFact converts context to a CLIPS fact
func (e *ClipsRuleEngine) contextToFact(context *Context) map[string]interface{} {
	keyRoot := "unknown"
	keyMode := "unknown"

	if context.KeySignature != nil {
		keyRoot = context.KeySignature.Root.String()
		switch context.KeySignature.Mode {
		case Major:
			keyMode = "major"
		case Minor:
			keyMode = "minor"
		default:
			keyMode = "other"
		}
	}

	return map[string]interface{}{
		"template": "context",
		"values": map[string]interface{}{
			"key-root":          map[string]interface{}{"symbol": keyRoot},
			"key-mode":          map[string]interface{}{"symbol": keyMode},
			"time-num":          context.TimeSignature.Numerator,
			"time-denom":        context.TimeSignature.Denominator,
			"tempo":             context.Tempo,
			"ticks-per-quarter": context.TicksPerQuarter,
		},
	}
}

// noteToFact converts a note to a CLIPS fact
func (e *ClipsRuleEngine) noteToFact(note *Note, index int) map[string]interface{} {
	pc := PitchClassFromMidi(note.Pitch)
	octave := int(note.Pitch)/12 - 1

	return map[string]interface{}{
		"template": "note",
		"values": map[string]interface{}{
			"index":       index,
			"pitch":       note.Pitch,
			"pitch-class": map[string]interface{}{"symbol": pc.String()},
			"octave":      octave,
			"duration":    note.Duration,
			"velocity":    note.Velocity,
			"start-tick":  note.StartTick,
		},
	}
}

// intervalToFact converts an interval to a CLIPS fact
func (e *ClipsRuleEngine) intervalToFact(interval *IntervalInfo) map[string]interface{} {
	var quality string
	switch interval.Quality {
	case PerfectConsonance:
		quality = "perfect-consonance"
	case ImperfectConsonance:
		quality = "imperfect-consonance"
	case MildDissonance:
		quality = "mild-dissonance"
	case StrongDissonance:
		quality = "strong-dissonance"
	}

	var direction string
	switch interval.Direction {
	case Ascending:
		direction = "ascending"
	case Descending:
		direction = "descending"
	case Unison:
		direction = "unison"
	}

	return map[string]interface{}{
		"template": "interval",
		"values": map[string]interface{}{
			"from-index": interval.FromIndex,
			"to-index":   interval.ToIndex,
			"semitones":  interval.Semitones,
			"quality":    map[string]interface{}{"symbol": quality},
			"direction":  map[string]interface{}{"symbol": direction},
		},
	}
}

// ExtractAdjustments extracts scoring adjustments from CLIPS output
func (e *ClipsRuleEngine) ExtractAdjustments(output map[string]interface{}) []ScoringAdjustment {
	var adjustments []ScoringAdjustment

	conclusions, ok := output["conclusions"].([]interface{})
	if !ok {
		return adjustments
	}

	for _, conclusion := range conclusions {
		c, ok := conclusion.(map[string]interface{})
		if !ok {
			continue
		}

		template, _ := c["template"].(string)
		if template != "scoring-adjustment" {
			continue
		}

		values, ok := c["values"].(map[string]interface{})
		if !ok {
			continue
		}

		// Handle dimension which might be a symbol (object with "symbol" key) or a string
		dimension := "unknown"
		if d, ok := values["dimension"]; ok {
			if dm, ok := d.(map[string]interface{}); ok {
				if sym, ok := dm["symbol"].(string); ok {
					dimension = sym
				}
			} else if s, ok := d.(string); ok {
				dimension = s
			}
		}

		// Handle amount (the CLIPS template uses 'amount', not 'adjustment')
		adjustment := getFloat32OrDefault(values, "amount", 0)
		if adjustment == 0 {
			adjustment = getFloat32OrDefault(values, "adjustment", 0)
		}

		adj := ScoringAdjustment{
			Dimension:  dimension,
			Adjustment: adjustment,
			Reason:     getStringOrDefault(values, "reason", ""),
			RuleName:   getStringOrDefault(values, "rule-name", "unknown"),
		}
		adjustments = append(adjustments, adj)
	}

	return adjustments
}

// ExtractSuggestions extracts suggestions from CLIPS output
func (e *ClipsRuleEngine) ExtractSuggestions(output map[string]interface{}) []ClipsSuggestion {
	var suggestions []ClipsSuggestion

	conclusions, ok := output["conclusions"].([]interface{})
	if !ok {
		return suggestions
	}

	for _, conclusion := range conclusions {
		c, ok := conclusion.(map[string]interface{})
		if !ok {
			continue
		}

		template, _ := c["template"].(string)
		if template != "suggestion" {
			continue
		}

		values, ok := c["values"].(map[string]interface{})
		if !ok {
			continue
		}

		sug := ClipsSuggestion{
			Category: getStringOrDefault(values, "category", "general"),
			Severity: getStringOrDefault(values, "severity", "info"),
			Message:  getStringOrDefault(values, "message", ""),
			RuleName: getStringOrDefault(values, "rule-name", "unknown"),
		}

		// Extract note indices if present
		if indices, ok := values["note-indices"].([]interface{}); ok {
			for _, idx := range indices {
				if n, ok := idx.(float64); ok {
					sug.NoteIndices = append(sug.NoteIndices, int(n))
				}
			}
		}

		suggestions = append(suggestions, sug)
	}

	return suggestions
}

// Analyze runs CLIPS rules on a sequence and returns results (synchronous wrapper)
func (e *ClipsRuleEngine) Analyze(sequence *Sequence) (*ClipsResult, error) {
	return e.AnalyzeContext(context.Background(), sequence)
}

// AnalyzeContext runs CLIPS rules on a sequence with context support
func (e *ClipsRuleEngine) AnalyzeContext(ctx context.Context, sequence *Sequence) (*ClipsResult, error) {
	start := time.Now()

	if !e.available {
		return &ClipsResult{
			Adjustments:     []ScoringAdjustment{},
			Suggestions:     []ClipsSuggestion{},
			RulesFired:      0,
			ExecutionTimeMs: uint64(time.Since(start).Milliseconds()),
		}, nil
	}

	// Convert sequence to facts
	factsJSON := e.SequenceToFacts(sequence)

	// Log facts for debugging if CLIPS_DEBUG is set
	if os.Getenv("CLIPS_DEBUG") != "" {
		data, _ := json.MarshalIndent(factsJSON, "", "  ")
		fmt.Fprintf(os.Stderr, "CLIPS Facts:\n%s\n", string(data))
	}

	// Create CLIPS provider (FFI-backed)
	provider, err := nxuskit.NewClipsFFIProvider(e.rulesDir)
	if err != nil {
		return nil, fmt.Errorf("failed to create CLIPS provider: %w", err)
	}

	// Build the request - model field specifies which rule files to load
	ruleFiles := "templates.clp,music-theory.clp,scoring-adjustments.clp,suggestions.clp"
	factsData, _ := json.Marshal(factsJSON)

	req := &nxuskit.ChatRequest{
		Model: ruleFiles,
		Messages: []nxuskit.Message{
			nxuskit.UserMessage(string(factsData)),
		},
	}

	// Execute CLIPS rules
	resp, err := provider.Chat(ctx, req)
	if err != nil {
		// Check if it's a license error - if so, fall back gracefully
		if errors.Is(err, nxuskit.ErrLicenseRequired) {
			if os.Getenv("CLIPS_DEBUG") != "" {
				fmt.Fprintf(os.Stderr, "CLIPS: License required, using deterministic scoring\n")
			}
			return &ClipsResult{
				Adjustments:     []ScoringAdjustment{},
				Suggestions:     []ClipsSuggestion{},
				RulesFired:      0,
				ExecutionTimeMs: uint64(time.Since(start).Milliseconds()),
			}, nil
		}
		return nil, fmt.Errorf("CLIPS execution failed: %w", err)
	}

	executionTimeMs := uint64(time.Since(start).Milliseconds())

	// Parse the response
	var output map[string]interface{}
	if err := json.Unmarshal([]byte(resp.Content), &output); err != nil {
		return nil, fmt.Errorf("failed to parse CLIPS output: %w", err)
	}

	// Extract adjustments and suggestions
	adjustments := e.ExtractAdjustments(output)
	suggestions := e.ExtractSuggestions(output)

	// Get rules fired count
	var rulesFired uint32
	if stats, ok := output["stats"].(map[string]interface{}); ok {
		if rf, ok := stats["total_rules_fired"].(float64); ok {
			rulesFired = uint32(rf)
		}
	}

	return &ClipsResult{
		Adjustments:     adjustments,
		Suggestions:     suggestions,
		RulesFired:      rulesFired,
		ExecutionTimeMs: executionTimeMs,
	}, nil
}

// ApplyAdjustments applies scoring adjustments to dimension scores
func ApplyAdjustments(scores map[string]float32, adjustments []ScoringAdjustment) {
	for _, adj := range adjustments {
		if score, ok := scores[adj.Dimension]; ok {
			newScore := score + adj.Adjustment
			if newScore < 0 {
				newScore = 0
			} else if newScore > 100 {
				newScore = 100
			}
			scores[adj.Dimension] = newScore
		}
	}
}

// Helper functions

func getStringOrDefault(m map[string]interface{}, key, defaultVal string) string {
	if v, ok := m[key].(string); ok {
		return v
	}
	return defaultVal
}

func getFloat32OrDefault(m map[string]interface{}, key string, defaultVal float32) float32 {
	if v, ok := m[key].(float64); ok {
		return float32(v)
	}
	return defaultVal
}

// LoadRulesDir returns the default rules directory path
func LoadRulesDir() string {
	// Check environment variable first
	if dir := os.Getenv("RIFFER_RULES_DIR"); dir != "" {
		return dir
	}

	// Try relative to executable
	if exe, err := os.Executable(); err == nil {
		dir := filepath.Join(filepath.Dir(exe), "rules")
		if info, err := os.Stat(dir); err == nil && info.IsDir() {
			return dir
		}
	}

	// Default to current directory
	return "./rules"
}

File: examples/apps/riffer/rust/engine/clips_bridge.rs
//! CLIPS Rule Engine Bridge for Riffer
//!
//! Provides integration with the CLIPS expert system for:
//! - Scoring adjustments (bonuses/penalties based on music theory rules)
//! - Context-aware suggestions for improvement
//!
//! Uses nxuskit's NxuskitProvider with provider_type "clips" for rule execution.
//!
//! Note: This module uses NxuskitProvider for CLIPS integration.
//! Currently unused but available for future activation.

use nxuskit_examples_clips_wire::{ClipsInputWire, ClipsOutputWire};
use serde::{Deserialize, Serialize};
use std::path::Path;

use crate::errors::Result;
#[allow(unused_imports)]
use crate::errors::RifferError;
use crate::theory::intervals::{IntervalInfo, analyze_intervals};
use crate::theory::scales::is_in_scale;
use crate::types::{Context, IntervalQuality, Mode, Note, Sequence};

/// Scoring adjustment from CLIPS rules
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScoringAdjustment {
    /// Which dimension this adjustment applies to
    pub dimension: String,
    /// Adjustment value (positive = bonus, negative = penalty)
    pub adjustment: f32,
    /// Reason for the adjustment
    pub reason: String,
    /// Rule that triggered this adjustment
    pub rule_name: String,
}

/// Suggestion from CLIPS rules
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClipsSuggestion {
    /// Suggestion category (harmony, melody, rhythm, dynamics, resolution, structure)
    pub category: String,
    /// Severity level (info, suggestion, warning)
    pub severity: String,
    /// The suggestion text
    pub message: String,
    /// Rule that generated this suggestion
    pub rule_name: String,
    /// Optional note indices this applies to
    pub note_indices: Option<Vec<usize>>,
}

/// Result from CLIPS rule execution
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ClipsResult {
    /// Scoring adjustments (bonuses/penalties)
    pub adjustments: Vec<ScoringAdjustment>,
    /// Improvement suggestions
    pub suggestions: Vec<ClipsSuggestion>,
    /// Number of rules that fired
    pub rules_fired: u32,
    /// Execution time in milliseconds
    pub execution_time_ms: u64,
}

/// CLIPS Rule Engine for music analysis
pub struct ClipsRuleEngine {
    /// Path to the rules directory
    #[allow(dead_code)]
    rules_dir: std::path::PathBuf,
    /// Whether CLIPS is available
    available: bool,
}

impl ClipsRuleEngine {
    /// Create a new CLIPS rule engine
    ///
    /// # Arguments
    /// * `rules_dir` - Path to directory containing .clp rule files
    pub fn new(rules_dir: &Path) -> Result<Self> {
        let rules_dir = rules_dir.to_path_buf();

        // Check if rules directory exists
        let available = rules_dir.exists() && rules_dir.is_dir();

        if !available {
            eprintln!(
                "Warning: CLIPS rules directory not found at {}. Using deterministic scoring only.",
                rules_dir.display()
            );
        }

        Ok(Self {
            rules_dir,
            available,
        })
    }

    /// Check if CLIPS is available
    pub fn is_available(&self) -> bool {
        self.available
    }

    /// Convert a sequence to CLIPS provider-chat input (`ClipsInput`-shaped JSON).
    pub fn sequence_to_facts(
        &self,
        sequence: &Sequence,
    ) -> std::result::Result<ClipsInputWire, serde_json::Error> {
        let mut facts: Vec<serde_json::Value> = Vec::new();

        // Add context fact
        let context_fact = self.context_to_fact(&sequence.context);
        facts.push(context_fact);

        // Add note facts
        for (idx, note) in sequence.notes.iter().enumerate() {
            let note_fact = self.note_to_fact(note, idx);
            facts.push(note_fact);
        }

        // Add interval facts
        let intervals = analyze_intervals(&sequence.notes);
        for interval in &intervals {
            let interval_fact = self.interval_to_fact(interval);
            facts.push(interval_fact);
        }

        // Add scale membership facts
        if let Some(key) = &sequence.context.key_signature {
            for (idx, note) in sequence.notes.iter().enumerate() {
                let pc = note.pitch_class();
                let in_scale = is_in_scale(pc, key);
                let membership_fact = serde_json::json!({
                    "template": "scale-membership",
                    "values": {
                        "note-index": idx,
                        "pitch-class": {"symbol": pc.to_string()},
                        "in-scale": {"symbol": if in_scale { "yes" } else { "no" }}
                    }
                });
                facts.push(membership_fact);
            }
        }

        // Add rhythmic-entropy fact
        let unique_durations: std::collections::HashSet<_> =
            sequence.notes.iter().map(|n| n.duration).collect();
        let rhythmic_variety = if sequence.notes.is_empty() {
            0.0
        } else {
            (unique_durations.len() as f32 / sequence.notes.len() as f32).min(1.0)
        };
        let rhythmic_entropy_fact = serde_json::json!({
            "template": "rhythmic-entropy",
            "values": {
                "value": rhythmic_variety,
                "unique-durations": unique_durations.len(),
                "total-notes": sequence.notes.len()
            }
        });
        facts.push(rhythmic_entropy_fact);

        // Add dynamics-summary fact
        if !sequence.notes.is_empty() {
            let velocities: Vec<u8> = sequence.notes.iter().map(|n| n.velocity).collect();
            let min_vel = *velocities.iter().min().unwrap_or(&0);
            let max_vel = *velocities.iter().max().unwrap_or(&127);
            let velocity_range = max_vel.saturating_sub(min_vel);

            // Calculate standard deviation
            let mean = velocities.iter().map(|&v| v as f32).sum::<f32>() / velocities.len() as f32;
            let variance = velocities
                .iter()
                .map(|&v| (v as f32 - mean).powi(2))
                .sum::<f32>()
                / velocities.len() as f32;
            let velocity_std = variance.sqrt();

            let dynamics_fact = serde_json::json!({
                "template": "dynamics-summary",
                "values": {
                    "min-velocity": min_vel,
                    "max-velocity": max_vel,
                    "velocity-range": velocity_range,
                    "velocity-std": velocity_std
                }
            });
            facts.push(dynamics_fact);
        }

        // Add contour fact
        if !intervals.is_empty() {
            let mut direction_changes = 0;
            let mut ascending_count = 0;
            let mut descending_count = 0;
            let mut last_direction: Option<bool> = None;

            for interval in &intervals {
                let is_ascending = interval.semitones > 0;
                let is_descending = interval.semitones < 0;

                if is_ascending {
                    ascending_count += 1;
                } else if is_descending {
                    descending_count += 1;
                }

                if let Some(was_ascending) = last_direction {
                    if (was_ascending && is_descending) || (!was_ascending && is_ascending) {
                        direction_changes += 1;
                    }
                }
                if is_ascending || is_descending {
                    last_direction = Some(is_ascending);
                }
            }

            let contour_type = if direction_changes == 0 && ascending_count > descending_count {
                "ascending"
            } else if direction_changes == 0 && descending_count > ascending_count {
                "descending"
            } else if ascending_count == descending_count {
                "static"
            } else if direction_changes >= 2 {
                "wave"
            } else if ascending_count > descending_count {
                "arch"
            } else {
                "inverse-arch"
            };

            let contour_fact = serde_json::json!({
                "template": "contour",
                "values": {
                    "type": {"symbol": contour_type},
                    "direction-changes": direction_changes
                }
            });
            facts.push(contour_fact);
        }

        serde_json::from_value(serde_json::json!({
            "facts": facts,
            "config": {
                "include_trace": true,
                "max_rules": 100
            }
        }))
    }

    /// Convert context to a CLIPS fact
    fn context_to_fact(&self, context: &Context) -> serde_json::Value {
        let detected_key = context
            .key_signature
            .as_ref()
            .map(|k| k.root.to_string())
            .unwrap_or_else(|| "C".to_string());

        let detected_mode = context
            .key_signature
            .as_ref()
            .map(|k| match k.mode {
                Mode::Major => "major",
                Mode::Minor => "minor",
                Mode::Dorian => "dorian",
                Mode::Phrygian => "phrygian",
                Mode::Lydian => "lydian",
                Mode::Mixolydian => "mixolydian",
                Mode::Aeolian => "aeolian",
                Mode::Locrian => "locrian",
            })
            .unwrap_or("major");

        serde_json::json!({
            "template": "context",
            "values": {
                "detected-key": detected_key,
                "detected-mode": {"symbol": detected_mode},
                "confidence": 0.8,
                "tempo": context.tempo,
                "ticks-per-quarter": context.ticks_per_quarter
            }
        })
    }

    /// Convert a note to a CLIPS fact
    fn note_to_fact(&self, note: &Note, index: usize) -> serde_json::Value {
        let pc = note.pitch_class();
        serde_json::json!({
            "template": "note",
            "values": {
                "index": index,
                "pitch": note.pitch,
                "pitch-class": {"symbol": pc.to_string()},
                "octave": note.octave(),
                "duration": note.duration,
                "velocity": note.velocity,
                "start-tick": note.start_tick
            }
        })
    }

    /// Convert an interval to a CLIPS fact
    fn interval_to_fact(&self, interval: &IntervalInfo) -> serde_json::Value {
        let quality = match interval.quality {
            IntervalQuality::PerfectConsonance => "perfect-consonance",
            IntervalQuality::ImperfectConsonance => "imperfect-consonance",
            IntervalQuality::MildDissonance => "mild-dissonance",
            IntervalQuality::StrongDissonance => "strong-dissonance",
        };

        // Generate a human-readable interval name
        let name = crate::theory::intervals::interval_name(interval.semitones as i8);

        serde_json::json!({
            "template": "interval",
            "values": {
                "from-index": interval.from_index,
                "to-index": interval.to_index,
                "semitones": interval.semitones,
                "name": name,
                "quality": {"symbol": quality}
            }
        })
    }

    /// Extract scoring adjustments from CLIPS output
    pub fn extract_adjustments(&self, output: &serde_json::Value) -> Vec<ScoringAdjustment> {
        let mut adjustments = Vec::new();

        if let Some(conclusions) = output.get("conclusions").and_then(|c| c.as_array()) {
            for conclusion in conclusions {
                if conclusion.get("template").and_then(|t| t.as_str()) == Some("scoring-adjustment")
                {
                    if let Some(values) = conclusion.get("values") {
                        // Handle dimension which might be a symbol (object with "symbol" key) or a string
                        let dimension = values
                            .get("dimension")
                            .map(|d| {
                                // Try as object with symbol key first
                                if let Some(sym) = d.get("symbol").and_then(|s| s.as_str()) {
                                    sym.to_string()
                                } else if let Some(s) = d.as_str() {
                                    s.to_string()
                                } else {
                                    "unknown".to_string()
                                }
                            })
                            .unwrap_or_else(|| "unknown".to_string());

                        // Handle amount (the template uses 'amount', not 'adjustment')
                        let amount = values
                            .get("amount")
                            .and_then(|a| a.as_f64())
                            .or_else(|| values.get("adjustment").and_then(|a| a.as_f64()))
                            .unwrap_or(0.0) as f32;

                        let adjustment = ScoringAdjustment {
                            dimension,
                            adjustment: amount,
                            reason: values
                                .get("reason")
                                .and_then(|r| r.as_str())
                                .unwrap_or("")
                                .to_string(),
                            rule_name: values
                                .get("rule-name")
                                .and_then(|r| r.as_str())
                                .unwrap_or("unknown")
                                .to_string(),
                        };
                        adjustments.push(adjustment);
                    }
                }
            }
        }

        adjustments
    }

    /// Extract suggestions from CLIPS output
    pub fn extract_suggestions(&self, output: &serde_json::Value) -> Vec<ClipsSuggestion> {
        let mut suggestions = Vec::new();

        if let Some(conclusions) = output.get("conclusions").and_then(|c| c.as_array()) {
            for conclusion in conclusions {
                if conclusion.get("template").and_then(|t| t.as_str()) == Some("suggestion") {
                    if let Some(values) = conclusion.get("values") {
                        let suggestion = ClipsSuggestion {
                            category: values
                                .get("category")
                                .and_then(|c| c.as_str())
                                .unwrap_or("general")
                                .to_string(),
                            severity: values
                                .get("severity")
                                .and_then(|s| s.as_str())
                                .unwrap_or("info")
                                .to_string(),
                            message: values
                                .get("message")
                                .and_then(|m| m.as_str())
                                .unwrap_or("")
                                .to_string(),
                            rule_name: values
                                .get("rule-name")
                                .and_then(|r| r.as_str())
                                .unwrap_or("unknown")
                                .to_string(),
                            note_indices: values
                                .get("note-indices")
                                .and_then(|n| n.as_array())
                                .map(|arr| {
                                    arr.iter()
                                        .filter_map(|v| v.as_u64().map(|n| n as usize))
                                        .collect()
                                }),
                        };
                        suggestions.push(suggestion);
                    }
                }
            }
        }

        suggestions
    }

    /// Run CLIPS rules on a sequence and return results
    ///
    /// This is the main entry point for CLIPS integration.
    /// If CLIPS is not available, returns an empty result.
    pub async fn analyze(&self, sequence: &Sequence) -> Result<ClipsResult> {
        use nxuskit::{ChatRequest, Message, NxuskitProvider, ProviderConfig};

        if !self.available {
            return Ok(ClipsResult::default());
        }

        // Build provider via NxuskitProvider with clips provider type
        let provider = NxuskitProvider::new(ProviderConfig {
            provider_type: "clips".to_string(),
            model: Some(self.rules_dir.to_str().unwrap_or(".").to_string()),
            ..Default::default()
        })
        .map_err(|e| RifferError::ClipsError(e.to_string()))?;

        // Convert sequence to ClipsInput-shaped wire (shared crate mirrors engine JSON).
        let clips_input = self
            .sequence_to_facts(sequence)
            .map_err(|e| RifferError::ClipsError(format!("CLIPS input wire: {e}")))?;

        let input_json = serde_json::to_string(&clips_input)
            .map_err(|e| RifferError::ClipsError(format!("CLIPS input JSON: {e}")))?;

        // Create request with all rule files
        let rule_files = "templates.clp,music-theory.clp,scoring-adjustments.clp,suggestions.clp";
        let request = ChatRequest::new(rule_files).with_message(Message::user(input_json));

        // Execute rules
        let start = std::time::Instant::now();
        let response = provider
            .chat(request)
            .map_err(|e| RifferError::ClipsError(e.to_string()))?;
        let execution_time_ms = start.elapsed().as_millis() as u64;

        // Parse output
        let clips_output: ClipsOutputWire = serde_json::from_str(&response.content)
            .map_err(|e| RifferError::ClipsError(format!("Failed to parse CLIPS output: {}", e)))?;

        let output_val = serde_json::to_value(&clips_output)
            .map_err(|e| RifferError::ClipsError(format!("CLIPS output re-encode: {e}")))?;

        // Extract results (reuse Value-based helpers)
        let adjustments = self.extract_adjustments(&output_val);
        let suggestions = self.extract_suggestions(&output_val);
        let rules_fired = clips_output.stats.total_rules_fired as u32;

        Ok(ClipsResult {
            adjustments,
            suggestions,
            rules_fired,
            execution_time_ms,
        })
    }
}

/// Apply scoring adjustments to dimension scores
pub fn apply_adjustments(
    scores: &mut std::collections::HashMap<String, f32>,
    adjustments: &[ScoringAdjustment],
) {
    for adj in adjustments {
        if let Some(score) = scores.get_mut(&adj.dimension) {
            *score = (*score + adj.adjustment).clamp(0.0, 100.0);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_note_to_fact() {
        let engine = ClipsRuleEngine {
            rules_dir: std::path::PathBuf::from("."),
            available: true,
        };

        let note = Note::new(60, 480, 80, 0); // Middle C
        let fact = engine.note_to_fact(&note, 0);

        assert_eq!(fact["template"], "note");
        assert_eq!(fact["values"]["pitch"], 60);
        assert_eq!(fact["values"]["index"], 0);
    }

    #[test]
    fn test_extract_adjustments() {
        let engine = ClipsRuleEngine {
            rules_dir: std::path::PathBuf::from("."),
            available: true,
        };

        let output = serde_json::json!({
            "conclusions": [
                {
                    "template": "scoring-adjustment",
                    "values": {
                        "dimension": "harmonic-coherence",
                        "adjustment": 5.0,
                        "reason": "Good use of leading tone",
                        "rule-name": "leading-tone-bonus"
                    }
                }
            ]
        });

        let adjustments = engine.extract_adjustments(&output);
        assert_eq!(adjustments.len(), 1);
        assert_eq!(adjustments[0].dimension, "harmonic-coherence");
        assert_eq!(adjustments[0].adjustment, 5.0);
    }

    #[test]
    fn test_extract_suggestions() {
        let engine = ClipsRuleEngine {
            rules_dir: std::path::PathBuf::from("."),
            available: true,
        };

        let output = serde_json::json!({
            "conclusions": [
                {
                    "template": "suggestion",
                    "values": {
                        "category": "resolution",
                        "severity": "warning",
                        "message": "Unresolved tritone at measure 4",
                        "rule-name": "unresolved-tritone"
                    }
                }
            ]
        });

        let suggestions = engine.extract_suggestions(&output);
        assert_eq!(suggestions.len(), 1);
        assert_eq!(suggestions[0].category, "resolution");
        assert_eq!(suggestions[0].severity, "warning");
    }
}
